File: backends/go/keyframe_expression_test.go
package main

import (
	"math"
	"strconv"
	"strings"
	"testing"
)

// Minimal ffmpeg-expression evaluator for tests. Supports the exact
// subset our compiler emits: lt, gte, pow, if, +-*/, parens, decimal
// numbers, and `t`. Mirrors the JS helper in the TS backend so the
// same assertions hold both sides.

type exprNode interface{ eval(t float64) float64 }
type numNode struct{ v float64 }
type tNode struct{}
type binNode struct {
	op       byte
	lhs, rhs exprNode
}
type callNode struct {
	name string
	args []exprNode
}

func (n numNode) eval(_ float64) float64 { return n.v }
func (tNode) eval(t float64) float64     { return t }
func (n binNode) eval(t float64) float64 {
	a, b := n.lhs.eval(t), n.rhs.eval(t)
	switch n.op {
	case '+':
		return a + b
	case '-':
		return a - b
	case '*':
		return a * b
	case '/':
		return a / b
	}
	panic("bad op")
}
func (n callNode) eval(t float64) float64 {
	switch n.name {
	case "lt":
		if n.args[0].eval(t) < n.args[1].eval(t) {
			return 1
		}
		return 0
	case "gte":
		if n.args[0].eval(t) >= n.args[1].eval(t) {
			return 1
		}
		return 0
	case "pow":
		return math.Pow(n.args[0].eval(t), n.args[1].eval(t))
	case "if":
		if n.args[0].eval(t) != 0 {
			return n.args[1].eval(t)
		}
		return n.args[2].eval(t)
	}
	panic("unknown fn " + n.name)
}

type parser struct {
	src string
	pos int
}

func (p *parser) peek() byte {
	if p.pos >= len(p.src) {
		return 0
	}
	return p.src[p.pos]
}
func (p *parser) take() byte { c := p.peek(); p.pos++; return c }

func (p *parser) parseExpr() exprNode { return p.parseAddSub() }
func (p *parser) parseAddSub() exprNode {
	lhs := p.parseMulDiv()
	for p.peek() == '+' || p.peek() == '-' {
		op := p.take()
		lhs = binNode{op, lhs, p.parseMulDiv()}
	}
	return lhs
}
func (p *parser) parseMulDiv() exprNode {
	lhs := p.parseUnary()
	for p.peek() == '*' || p.peek() == '/' {
		op := p.take()
		lhs = binNode{op, lhs, p.parseUnary()}
	}
	return lhs
}
func (p *parser) parseUnary() exprNode {
	if p.peek() == '-' {
		p.take()
		return binNode{'-', numNode{0}, p.parseAtom()}
	}
	if p.peek() == '+' {
		p.take()
	}
	return p.parseAtom()
}
func (p *parser) parseAtom() exprNode {
	c := p.peek()
	if c == '(' {
		p.take()
		e := p.parseExpr()
		if p.peek() != ')' {
			panic("expected ) at " + strconv.Itoa(p.pos))
		}
		p.take()
		return e
	}
	if (c >= '0' && c <= '9') || c == '.' {
		start := p.pos
		for {
			c := p.peek()
			if (c >= '0' && c <= '9') || c == '.' {
				p.take()
			} else {
				break
			}
		}
		v, err := strconv.ParseFloat(p.src[start:p.pos], 64)
		if err != nil {
			panic(err)
		}
		return numNode{v}
	}
	if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') {
		start := p.pos
		for {
			c := p.peek()
			if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') {
				p.take()
			} else {
				break
			}
		}
		name := p.src[start:p.pos]
		if p.peek() == '(' {
			p.take()
			var args []exprNode
			if p.peek() != ')' {
				args = append(args, p.parseExpr())
				for p.peek() == ',' {
					p.take()
					args = append(args, p.parseExpr())
				}
			}
			if p.peek() != ')' {
				panic("expected ) after fn args")
			}
			p.take()
			return callNode{name, args}
		}
		if name == "t" {
			return tNode{}
		}
		panic("unknown identifier " + name)
	}
	panic("unexpected char " + string(c))
}

func evalExpr(expr string, t float64) float64 {
	p := &parser{src: strings.ReplaceAll(expr, " ", "")}
	return p.parseExpr().eval(t)
}

// ----- compiler tests --------------------------------------------------

func TestCompileKeyframeExpression_NoKf_ReturnsFallback(t *testing.T) {
	if got := compileKeyframeExpression(nil, "panX", 0); got != "0" {
		t.Fatalf("got %q", got)
	}
	if got := compileKeyframeExpression([]Keyframe{}, "scale", 1.5); got != "1.5" {
		t.Fatalf("got %q", got)
	}
}

func TestCompileKeyframeExpression_SingleKf_Constant(t *testing.T) {
	kfs := []Keyframe{{ID: "k1", Prop: "scale", Time: 1000, Value: 2}}
	if got := compileKeyframeExpression(kfs, "scale", 1); got != "2" {
		t.Fatalf("got %q", got)
	}
}

func TestCompileKeyframeExpression_Linear(t *testing.T) {
	kfs := []Keyframe{
		{ID: "a", Prop: "scale", Time: 0, Value: 1},
		{ID: "b", Prop: "scale", Time: 1000, Value: 2},
	}
	expr := compileKeyframeExpression(kfs, "scale", 1)
	cases := []struct{ t, want float64 }{
		{0, 1}, {0.5, 1.5}, {1, 2}, {2, 2}, // after-last held
	}
	for _, c := range cases {
		got := evalExpr(expr, c.t)
		if math.Abs(got-c.want) > 1e-5 {
			t.Errorf("t=%v want %v got %v\nexpr=%s", c.t, c.want, got, expr)
		}
	}
}

func TestCompileKeyframeExpression_EaseIn(t *testing.T) {
	kfs := []Keyframe{
		{ID: "a", Prop: "scale", Time: 0, Value: 0, Easing: "easeIn"},
		{ID: "b", Prop: "scale", Time: 1000, Value: 1},
	}
	expr := compileKeyframeExpression(kfs, "scale", 0)
	// easeIn(0.5) = 0.5^3 = 0.125
	if got := evalExpr(expr, 0.5); math.Abs(got-0.125) > 1e-5 {
		t.Fatalf("want 0.125 got %v\nexpr=%s", got, expr)
	}
}

func TestCompileKeyframeExpression_EaseOut(t *testing.T) {
	kfs := []Keyframe{
		{ID: "a", Prop: "scale", Time: 0, Value: 0, Easing: "easeOut"},
		{ID: "b", Prop: "scale", Time: 1000, Value: 1},
	}
	expr := compileKeyframeExpression(kfs, "scale", 0)
	// easeOut(0.5) = 1 - (1-0.5)^3 = 1 - 0.125 = 0.875
	if got := evalExpr(expr, 0.5); math.Abs(got-0.875) > 1e-5 {
		t.Fatalf("want 0.875 got %v\nexpr=%s", got, expr)
	}
}

func TestCompileKeyframeExpression_EaseInOut_Symmetric(t *testing.T) {
	kfs := []Keyframe{
		{ID: "a", Prop: "scale", Time: 0, Value: 0, Easing: "easeInOut"},
		{ID: "b", Prop: "scale", Time: 1000, Value: 1},
	}
	expr := compileKeyframeExpression(kfs, "scale", 0)
	if got := evalExpr(expr, 0.5); math.Abs(got-0.5) > 1e-5 {
		t.Fatalf("want 0.5 got %v\nexpr=%s", got, expr)
	}
}

func TestCompileKeyframeExpression_BeforeAndAfterHold(t *testing.T) {
	kfs := []Keyframe{
		{ID: "a", Prop: "panX", Time: 1000, Value: 50},
		{ID: "b", Prop: "panX", Time: 2000, Value: 100},
	}
	expr := compileKeyframeExpression(kfs, "panX", 0)
	if v := evalExpr(expr, 0); math.Abs(v-50) > 1e-5 {
		t.Errorf("before-first want 50 got %v\nexpr=%s", v, expr)
	}
	if v := evalExpr(expr, 1.5); math.Abs(v-75) > 1e-5 {
		t.Errorf("mid want 75 got %v\nexpr=%s", v, expr)
	}
	if v := evalExpr(expr, 3); math.Abs(v-100) > 1e-5 {
		t.Errorf("after-last want 100 got %v\nexpr=%s", v, expr)
	}
}

func TestCompileKeyframeExpression_IgnoresOtherProps(t *testing.T) {
	kfs := []Keyframe{{ID: "a", Prop: "panX", Time: 0, Value: 999}}
	if got := compileKeyframeExpression(kfs, "panY", 42); got != "42" {
		t.Fatalf("got %q", got)
	}
}

func TestCompileKeyframeExpression_ZeroLengthSegmentSkipped(t *testing.T) {
	kfs := []Keyframe{
		{ID: "a", Prop: "scale", Time: 1000, Value: 1},
		{ID: "b", Prop: "scale", Time: 1000, Value: 2}, // same time
		{ID: "c", Prop: "scale", Time: 2000, Value: 3},
	}
	expr := compileKeyframeExpression(kfs, "scale", 1)
	if strings.Contains(expr, "NaN") || strings.Contains(expr, "Infinity") {
		t.Fatalf("unsafe expr: %s", expr)
	}
	if v := evalExpr(expr, 1.5); math.Abs(v-2.5) > 1e-5 {
		t.Errorf("want 2.5 got %v\nexpr=%s", v, expr)
	}
}

func TestFormatNumber_IntegerStaysInteger(t *testing.T) {
	if got := formatNumber(2); got != "2" {
		t.Errorf("got %q", got)
	}
	if got := formatNumber(2.5); got != "2.5" {
		t.Errorf("got %q", got)
	}
	if got := formatNumber(0.1); got != "0.1" {
		t.Errorf("got %q", got)
	}
}

File: backends/ts/src/keyframe-expression.test.ts
import { describe, expect, it } from "vitest";
import type { Keyframe } from "@aicut/core";
import {
  compileKeyframeExpression,
  easingExpression,
} from "./keyframe-expression.js";

/**
 * The frontend's reference implementation — same math as
 * keyframes/interpolate.ts:applyEasing. We rebuild a JS evaluator
 * that mirrors ffmpeg's grammar narrowly (just lt/gte/pow/if) so
 * the compiled expression can be ROUND-TRIPPED back through JS and
 * compared against the frontend's interpolateProp output. That's
 * what proves "what ffmpeg renders == what the preview shows."
 */
function evalFfmpegExpr(expr: string, t: number): number {
  // Replace ffmpeg builtins with JS equivalents.
  const js = expr
    .replace(/\bif\(/g, "_if(")
    .replace(/\bgte\(/g, "_gte(")
    .replace(/\blt\(/g, "_lt(")
    .replace(/\bpow\(/g, "Math.pow(");
  const _if = (cond: number, a: number, b: number): number => (cond ? a : b);
  const _gte = (a: number, b: number): number => (a >= b ? 1 : 0);
  const _lt = (a: number, b: number): number => (a < b ? 1 : 0);
  // eslint-disable-next-line no-new-func
  const fn = new Function("t", "_if", "_gte", "_lt", `return ${js};`);
  return fn(t, _if, _gte, _lt) as number;
}

function kf(
  prop: "panX" | "panY" | "scale",
  time: number,
  value: number,
  easing?: "linear" | "easeIn" | "easeOut" | "easeInOut",
): Keyframe {
  const k: Keyframe = { id: `k-${prop}-${time}`, prop, time, value };
  if (easing) k.easing = easing;
  return k;
}

describe("compileKeyframeExpression", () => {
  it("no keyframes → returns the fallback as a literal", () => {
    expect(compileKeyframeExpression(undefined, "panX", 0)).toBe("0");
    expect(compileKeyframeExpression([], "scale", 1.5)).toBe("1.5");
    // Decimal trimming — no trailing zeros, no stranded dot.
    expect(compileKeyframeExpression([], "panX", 42.500000)).toBe("42.5");
  });

  it("single keyframe → its value as a literal (constant)", () => {
    const expr = compileKeyframeExpression(
      [kf("scale", 1000, 2)],
      "scale",
      1,
    );
    expect(expr).toBe("2");
    // At any t, value is constant.
    expect(evalFfmpegExpr(expr, 0)).toBe(2);
    expect(evalFfmpegExpr(expr, 5)).toBe(2);
  });

  it("ignores keyframes for other props", () => {
    const expr = compileKeyframeExpression(
      [kf("panX", 0, 100), kf("scale", 1000, 2)],
      "panY",
      50,
    );
    expect(expr).toBe("50");
  });

  it("piecewise linear: matches the frontend interpolator at sample t", () => {
    // Two kfs: scale 1 → 2 over 0..1000ms.
    const kfs = [kf("scale", 0, 1), kf("scale", 1000, 2)];
    const expr = compileKeyframeExpression(kfs, "scale", 1);
    expect(evalFfmpegExpr(expr, 0)).toBeCloseTo(1, 5);
    expect(evalFfmpegExpr(expr, 0.5)).toBeCloseTo(1.5, 5);
    expect(evalFfmpegExpr(expr, 1)).toBeCloseTo(2, 5);
    // After last → held.
    expect(evalFfmpegExpr(expr, 2)).toBeCloseTo(2, 5);
  });

  it("before first / after last → held at the boundary kf's value", () => {
    const kfs = [kf("panX", 1000, 50), kf("panX", 2000, 100)];
    const expr = compileKeyframeExpression(kfs, "panX", 0);
    expect(evalFfmpegExpr(expr, 0)).toBeCloseTo(50, 5); // held at first
    expect(evalFfmpegExpr(expr, 1.5)).toBeCloseTo(75, 5); // mid-segment
    expect(evalFfmpegExpr(expr, 3)).toBeCloseTo(100, 5); // held at last
  });

  it("easing — easeIn moves below linear in the middle", () => {
    const kfs = [kf("scale", 0, 0, "easeIn"), kf("scale", 1000, 1)];
    const expr = compileKeyframeExpression(kfs, "scale", 0);
    // Linear would be 0.5 at t=0.5; easeIn = 0.5³ = 0.125
    expect(evalFfmpegExpr(expr, 0.5)).toBeCloseTo(0.125, 5);
    // Endpoints still exact.
    expect(evalFfmpegExpr(expr, 0)).toBeCloseTo(0, 5);
    expect(evalFfmpegExpr(expr, 1)).toBeCloseTo(1, 5);
  });

  it("easing — easeOut moves above linear in the middle", () => {
    const kfs = [kf("scale", 0, 0, "easeOut"), kf("scale", 1000, 1)];
    const expr = compileKeyframeExpression(kfs, "scale", 0);
    // easeOut at t=0.5 → 1 - (1-0.5)³ = 1 - 0.125 = 0.875
    expect(evalFfmpegExpr(expr, 0.5)).toBeCloseTo(0.875, 5);
  });

  it("easing — easeInOut is symmetric around t=0.5", () => {
    const kfs = [kf("scale", 0, 0, "easeInOut"), kf("scale", 1000, 1)];
    const expr = compileKeyframeExpression(kfs, "scale", 0);
    expect(evalFfmpegExpr(expr, 0.5)).toBeCloseTo(0.5, 5);
  });

  it("multi-segment with per-segment easing", () => {
    // Linear seg 1 → easeOut seg 2.
    const kfs = [
      kf("scale", 0, 1),
      kf("scale", 1000, 2, "easeOut"),
      kf("scale", 2000, 3),
    ];
    const expr = compileKeyframeExpression(kfs, "scale", 1);
    // Seg 1 mid (t=0.5s) → linear: 1.5
    expect(evalFfmpegExpr(expr, 0.5)).toBeCloseTo(1.5, 5);
    // Seg 2 mid (t=1.5s) → easeOut: 2 + 0.875*(3-2) = 2.875
    expect(evalFfmpegExpr(expr, 1.5)).toBeCloseTo(2.875, 5);
  });

  it("zero-length segments are skipped without div-by-zero", () => {
    const kfs = [
      kf("scale", 1000, 1),
      kf("scale", 1000, 2), // exact dupe time
      kf("scale", 2000, 3),
    ];
    const expr = compileKeyframeExpression(kfs, "scale", 1);
    // Doesn't throw, doesn't contain "NaN" or "Infinity".
    expect(expr).not.toMatch(/NaN|Infinity/);
    // After the second kf @1s onward we ride the kf@1000=2 → kf@2000=3 ramp.
    expect(evalFfmpegExpr(expr, 1.5)).toBeCloseTo(2.5, 5);
  });

  it("ms → seconds conversion at 6-digit precision", () => {
    const kfs = [kf("scale", 1000, 1), kf("scale", 1500, 2)];
    const expr = compileKeyframeExpression(kfs, "scale", 1);
    // The expression should contain 1.5 (seconds), not 1500 (ms).
    expect(expr).toMatch(/1\.5/);
    expect(expr).not.toMatch(/1500/);
  });

  it("integer values stay integers — no '1.000000'", () => {
    const expr = compileKeyframeExpression(
      [kf("scale", 0, 1), kf("scale", 1000, 2)],
      "scale",
      1,
    );
    // No fractional zero noise — keeps the expr human-readable.
    expect(expr).not.toMatch(/\.0+\b/);
  });
});

describe("easingExpression — boundary identities", () => {
  it("rawT=0 / rawT=1 always evaluate to 0 / 1", () => {
    for (const kind of ["linear", "easeIn", "easeOut", "easeInOut"] as const) {
      const expr0 = easingExpression("0", kind);
      const expr1 = easingExpression("1", kind);
      expect(evalFfmpegExpr(expr0, 0)).toBeCloseTo(0, 6);
      expect(evalFfmpegExpr(expr1, 0)).toBeCloseTo(1, 6);
    }
  });
});
